### ISSUE TITLE
Check Rootid fails on Debian 8.4.0

### ISSUE THREAD

**Issue by fsundermeyer** [openSUSE/daps#360]
When running `daps ... validate` on DocBook 5 sources in Debian, DAPS hangs at

```
+ [[ -n book.sle.deployment ]]
++ /home/tux/daps/libexec/daps-xslt --stringparam rootid=book.sle.deployment --xinclude --stylesheet /home/tux/daps/daps-xslt/common/check_rootid.xsl --file /home/tux/doc-sle/xml/MAIN.SLEDS.xml /usr/bin/xsltproc
```

in lib/daps_functions.

The command that is called here is

```
/usr/bin/xsltproc  --xinclude  --stringparam rootid "book.sle.deployment"  /home/tux/daps/daps-xslt/common/check_rootid.xsl /home/tux/doc-sle/xml/MAIN.SLEDS.xml
```

Running the same command on a DocBook4 source file works.

Looks like something is wrong with the DB5 part of the stylesheet.

**Comment by tomschr** [comment]
Hmn... strange.

I've tested the `check_rootid.xsl` stylesheet with both versions, DB4 and DB5. Both show the expected result. However, I've tested it on openSUSE, not Debian.

I'm not sure why the result on Debian should be different. The stylesheet is a "normal" XSLT stylesheet, nothing fancy. It should work on Debian the same as for openSUSE. Otherwise there is a bug in the XSLT processor.

So I assume, it must be something different. Maybe there is a problematic version on Debian. Here is what I've used on my system:

```
$ xsltproc --version
Using libxml 20901, libxslt 10128 and libexslt 817
xsltproc was compiled against libxml 20901, libxslt 10128 and libexslt 817
libxslt 10128 was compiled against libxml 20901
libexslt 817 was compiled against libxml 20901
```

**Comment by fsundermeyer** [comment]
I know that it works on openSUSE - I would rather like to know why it does not work on Debian (and get this fixed). With the check rootid failing, DAPS is unusable on Debian.

**Comment by tomschr** [comment]
After some investigation, it seems it's a catalog problem (didn't we guess that?)

In the `entity-decl.ent` file, we have this entry:

``` xml
<!ENTITY % dbcent PUBLIC 
    "-//OASIS//ENTITIES DocBook Character Entities V4.5//EN" 
    "http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod">
%dbcent;
```

The `dbcentx.mod` pulls in all the entity files for DocBook. For example, this is one entry inside  `dbcentx.mod`:

``` xml
<!ENTITY % ISOamsa PUBLIC
"ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML"
"ent/isoamsa.ent">
```

The XML parser tries to resolve the publid identifier "ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML". As there is no catalog entry for it, it tries to use the system identifier. As it is retrieved from http://www.oasis-open.org/docbook/xml/4.5/, the absolute path for this entity file would be "http://www.oasis-open.org/docbook/xml/4.5/ent/isoamsa.ent". As there is also no catalog entry for this URL, the XML parser has no choice and fetches it from the URl.

That's the reason why it takes so long. The xsltproc didn't crash, it just do its job. ;-) You can see it if you use the `--load-trace` option for xsltproc. This gives you this output:

```
Loaded URL="http://www.oasis-open.org/docbook/xml/4.5/ent/isolat2.ent" ID="ISO 8879:1986//ENTITIES Added Latin 2//EN//XML"
```

whereas in openSUSE you can see this:

```
Loaded URL="file:/usr/share/xml/docbook/schema/dtd/4.5/ent/isolat2.ent" ID="ISO 8879:1986//ENTITIES Added Latin 2//EN//XML"
```

Solution: provide the missing catalog entries.

**Comment by tomschr** [comment]
Hmn, this is strange. When setting the environment variable `XML_DEBUG_CATALOG=1` this is the result (shortend):

#### openSUSE:

```
Resolve: pubID -//OASIS//ENTITIES DocBook Character Entities V4.5//EN sysID http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod
 Parsing catalog file:///etc/xml/catalog
file:///etc/xml/catalog added to file hash
file:/etc/xml/suse-catalog.xml not found in file hash
 Parsing catalog file:/etc/xml/suse-catalog.xml
file:/etc/xml/suse-catalog.xml added to file hash
Using rewriting rule http://www.oasis-open.org/docbook/xml/4.5
Resolve: pubID ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML sysID file:/usr/share/xml/docbook/schema/dtd/4.5/ent/isoamsa.ent
file:///usr/share/xmlcharent/xmlcharent.xml not found in file hash
 Parsing catalog file:///usr/share/xmlcharent/xmlcharent.xml
file:///usr/share/xmlcharent/xmlcharent.xml added to file hash
Trying public delegate file:///usr/share/xmlcharent/xmlcharent.xml
Found public match ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML
```

#### Debian

```
Resolve: pubID -//OASIS//ENTITIES DocBook Character Entities V4.5//EN sysID http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod
0 Parsing catalog file:///etc/xml/catalog
file:///etc/xml/catalog added to file hash
file:///etc/xml/docbook-xml.xml not found in file hash
0 Parsing catalog file:///etc/xml/docbook-xml.xml
file:///etc/xml/docbook-xml.xml added to file hash
Trying system delegate file:///etc/xml/docbook-xml.xml
Resolve URI http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod
Resolve: pubID ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML sysID http://www.oasis-open.org/docbook/xml/4.5/ent/isoamsa.ent
Trying system delegate file:///etc/xml/docbook-xml.xml
Resolve URI http://www.oasis-open.org/docbook/xml/4.5/ent/isoamsa.ent
Resolve: pubID ISO 8879:1986//ENTITIES Added Math Symbols: Binary Operators//EN//XML sysID http://www.oasis-open.org/docbook/xml/4.5/ent/isoamsb.ent
Trying system delegate file:///etc/xml/docbook-xml.xml
Resolve URI http://www.oasis-open.org/docbook/xml/4.5/ent/isoamsb.ent
```

**Comment by tomschr** [comment]
``` shell
$ XML_DEBUG_CATALOG=1 xmlcatalog /etc/xml/catalog "ISO 8879:1986//ENTITIES Publishing//EN//XML"
```

#### openSUSE 13.2

```
Resolve pubID ISO 8879:1986//ENTITIES Publishing//EN//XML
1532798720 Parsing catalog /etc/xml/catalog
/etc/xml/catalog added to file hash
file:///usr/share/xmlcharent/xmlcharent.xml not found in file hash
1532798720 Parsing catalog file:///usr/share/xmlcharent/xmlcharent.xml
file:///usr/share/xmlcharent/xmlcharent.xml added to file hash
Trying public delegate file:///usr/share/xmlcharent/xmlcharent.xml
Found public match ISO 8879:1986//ENTITIES Publishing//EN//XML
file:///usr/share/xmlcharent/entities/iso-pub.ent
Catalogs cleanup
Free catalog entry xmlcharent
Free catalog entry ISO 8879:1986//ENTITIES
Free catalog entry http://www.oasis-open.org/docbook/xmlcharent/0.3/
Free catalog entry suse-catalog.xml
Free catalog entry /etc/xml/catalog
Free catalog entry http://www.oasis-open.org/docbook/xmlcharent/0.3/
Free catalog entry ISO 8879:1986//ENTITIES Publishing//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Greek Letters//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Box and Line Drawing//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Greek Symbols//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Negated Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Numeric and Special Graphic//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Alternative Greek Symbols//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Diacritical Marks//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Monotoniko Greek//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Ordinary//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Russian Cyrillic//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES General Technical//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Delimiters//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Latin 1//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Binary Operators//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Latin 2//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Non-Russian Cyrillic//EN//XML
Free catalog entry file:///usr/share/xmlcharent/xmlcharent.xml
Free catalog entry
```

#### Debian (Jessie)

```
Resolve pubID ISO 8879:1986//ENTITIES Publishing//EN//XML
0 Parsing catalog file:///etc/xml/catalog
file:///etc/xml/catalog added to file hash
file:///etc/xml/sgml-data.xml not found in file hash
0 Parsing catalog file:///etc/xml/sgml-data.xml
file:///etc/xml/sgml-data.xml added to file hash
Trying public delegate file:///etc/xml/sgml-data.xml
file:///usr/share/xml/entities/xml-iso-entities-8879.1986/catalog.xml not found in file hash
0 Parsing catalog file:///usr/share/xml/entities/xml-iso-entities-8879.1986/catalog.xml
file:///usr/share/xml/entities/xml-iso-entities-8879.1986/catalog.xml added to file hash
Trying public delegate file:///usr/share/xml/entities/xml-iso-entities-8879.1986/catalog.xml
Found public match ISO 8879:1986//ENTITIES Publishing//EN//XML
file:/usr/share/xml/entities/xml-iso-entities-8879.1986/ISOpub.ent
Catalogs cleanup
Free catalog entry http://www.w3.org/Graphics/SVG/1.1/
Free catalog entry http://xml.ascc.net/xml/resource/qaml-xml
Free catalog entry http://www.w3.org/TR/2001/REC-SVG-20010904/
Free catalog entry /usr/share/sgml/dtd/svg
Free catalog entry -//W3C//DTD SVG
Free catalog entry +//ISBN 82-7640-023//DTD Frequently Asked Questions//EN
Free catalog entry /usr/share/sgml/dtd/qaml-xml.dtd
Free catalog entry ISO 8879:1986//ENTITIES
Free catalog entry +//IDN faq.org//DTD Frequently Asked Questions
Free catalog entry file:///etc/xml/sgml-data.xml
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Arrow Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Binary Operators//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Delimiters//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Negated Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Ordinary//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Math Symbols: Relations//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Box and Line Drawing//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Russian Cyrillic//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Non-Russian Cyrillic//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Diacritical Marks//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Greek Letters//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Monotoniko Greek//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Greek Symbols//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Alternative Greek Symbols//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Latin 1//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Added Latin 2//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Numeric and Special Graphic//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES Publishing//EN//XML
Free catalog entry ISO 8879:1986//ENTITIES General Technical//EN//XML
Free catalog entry file:///usr/share/xml/entities/xml-iso-entities-8879.1986/catalog.xml
Free catalog entry -//GlobalTransCorp//DTD XML Catalogs V1.0-Based Extension V1.0//EN
Free catalog entry ISO 8879:1986//ENTITIES
Free catalog entry -//Norman Walsh//DTD DocBook XML
Free catalog entry -//OASIS//DTD DocBook XML 5
Free catalog entry http://www.w3.org/Graphics/SVG/1.1/
Free catalog entry -//Norman Walsh//DTD CALS Table Model XML
Free catalog entry http://scrollkeeper.sourceforge.net/dtds/scrollkeeper-omf-1.0/
Free catalog entry http://docutils.sourceforge.net/
Free catalog entry http://www.oasis-open.org/committees/entity/release/1.0/catalog.dtd
Free catalog entry -//OASIS//DTD DocBook CALS Table Model
Free catalog entry -//OASIS//DTD XML Catalogs V1.0//EN
Free catalog entry http://www.w3.org/TR/2001/REC-SVG-20010904/
Free catalog entry http://docbook.sourceforge.net/release/xsl/
Free catalog entry /usr/share/sgml/dtd/svg
Free catalog entry -//Norman Walsh//ELEMENTS DocBk XML
Free catalog entry -//W3C//DTD SVG
Free catalog entry http://docbook.sourceforge.net/release/xsl-ns/
Free catalog entry http://www.lysator.liu.se/~alla/dia/
Free catalog entry http://www.oasis-open.org/docbook/xml/5
Free catalog entry http://docbook.sourceforge.net/release/xsl-ns/
Free catalog entry -//OASIS//DTD DocBook XML
Free catalog entry -//OASIS//DTD XML Exchange Table Model 19990315
Free catalog entry http://globaltranscorp.org/oasis/catalog/xml/tr9401.dtd
Free catalog entry http://docbook.org/xml/5
Free catalog entry http://docbook.sourceforge.net/release/xsl/
Free catalog entry +//IDN docutils.sourceforge.net/
Free catalog entry -//Norman Walsh//ENTITIES DocBk XML
Free catalog entry -//Normal Walsh//Exchange Table Model 19960430
Free catalog entry -//OASIS//ELEMENTS DocBook
Free catalog entry http://xml.ascc.net/xml/resource/qaml-xml
Free catalog entry http://www.lysator.liu.se/~alla/dia/dia-sheet-ns
Free catalog entry http://docbook.org/xml/
Free catalog entry -//OASIS//ENTITIES DocBook
Free catalog entry http://www.daa.com.au/~james/dia-shape-ns
Free catalog entry /usr/share/sgml/dtd/qaml-xml.dtd
Free catalog entry +//IDN faq.org//DTD Frequently Asked Questions
Free catalog entry +//ISBN 82-7640-023//DTD Frequently Asked Questions//EN
Free catalog entry http://www.oasis-open.org/docbook/xml/
Free catalog entry -//Norman Walsh//DTD DocBk XML
Free catalog entry file:///etc/xml/catalog
Free catalog entry
Free catalog entry
```

**Comment by tomschr** [comment]
The solution is quite simple:
1. add the following line to the `/etc/xml/docbook-xml.xml` file:
   
   ``` xml
   <delegateSystem systemIdStartString="http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod"
                   catalog="file:///usr/share/xml/docbook/schema/dtd/4.5/catalog.xml"/>
   ```
2.  add the following line to the `/usr/share/xml/docbook/schema/dtd/4.5/catalog.xml` file:
   
   ``` xml
   <system systemId="http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod"
           uri="dbcentx.mod"/>
   ```

If you want a separate catalog for DAPS, use only this snippet:

``` xml
<system systemId="http://www.oasis-open.org/docbook/xml/4.5/dbcentx.mod"
            uri="/usr/share/xml/docbook/schema/dtd/4.5/dbcentx.mod"/>
```

**Comment by fsundermeyer** [comment]
Fixed with merge commit f31c1572bd962
